### PR TITLE
Feature: (ISA-221) Include server-set CSS in GetFeatureInfo popup

### PR DESCRIPTION
**Pull request by JoshGx2000**
[ISA-221](https://jira.its.utas.edu.au/browse/ISA-221)

Thanks to the encapsulation provided by [Shadow-DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM), we can now include the style from GetFeatureInfo responses in addition to their body content. This has fixed the GFI display for a few layers.

Before:
Screen Shot 2022-08-05 at 5 22 23 pm
After:
Screen Shot 2022-08-05 at 5 22 37 pm

### CODE REVIEW

File: frontend/src/cljs/imas_seamap/map/utils.cljs [hunk 1]
```diff
@@ -198,19 +198,22 @@
...
 
 (defn feature-info-html
   [response]
   (let [parsed (.parseFromString (js/DOMParser.) response "text/html")
-        body  (first (array-seq (.querySelectorAll parsed "body")))]
-    (if (.-firstElementChild body)
-      {:info (.-innerHTML body)}
-      {:status :feature-info/empty})))
+        body  (first (array-seq (.querySelectorAll parsed "body")))
+        style  (first (array-seq (.querySelectorAll parsed "style")))] ; only grabs the first style element
```
Comment (markhepburn) [inline frontend/src/cljs/imas_seamap/map/utils.cljs:213]: Yeah I was going to look for this -- I suppose in the general instance it's not ideal, but for basic GFI responses that probably just have a single style block it's probably fine.

File: frontend/src/cljs/imas_seamap/map/utils.cljs [hunk 2]
```diff
@@ -219,13 +222,32 @@
...
-              "<table>" property-rows "</table>"
-              "</div>")}
-      {:status :feature-info/empty})))
+    (when (or id (not-empty properties))
+      {:style
+       (str
+        ".feature-info-json {"
+        "    max-height: 257px;"
+        "    overflow-y: scroll;"
+        "}"
```
Comment (markhepburn) [inline frontend/src/cljs/imas_seamap/map/utils.cljs:231]: Don't know why I didn't do it myself, but worth noting that clojure has multi-line strings out of the box.

File: frontend/src/cljs/imas_seamap/map/views.cljs [hunk 3]
```diff
@@ -202,7 +202,13 @@
...
                                    :icon  "error"}]
     ;; Default; we have actual content:
-    [:div {:dangerouslySetInnerHTML {:__html info-body}}]))
+    [:div
+     {:ref
+      (fn [element]
+        (when element
+          (set! (.-innerHTML element) nil)
+          (doseq [{:keys [_body _style] :as response} responses]
+            (.appendChild element (create-shadow-dom-element response)))))}]))
```
Comment (markhepburn) [inline frontend/src/cljs/imas_seamap/map/views.cljs:211]: What's the deal with refs, out of curiosity -- the only thing I'd want to ensure here is that the function isn't going to be called repeatedly (which I assume it doesn't, I've just never looked into it)

File: backend/catalogue/management/commands/generate_layer_previews.py [hunk 4]
```diff
@@ -10,8 +10,9 @@
 
 from catalogue.models import Layer
 
+Image.MAX_IMAGE_PIXELS = None
 
-basemap = Image.open(default_storage.open('basemap.png'))
+basemap = Image.open(default_storage.open('land_shallow_topo_21600.tif'))
```
Comment (markhepburn) [inline backend/catalogue/management/commands/generate_layer_previews.py:15]: Just an FYI, looks like this crept in.

**Review by markhepburn — APPROVED** [review]

File: frontend/src/cljs/imas_seamap/map/utils.cljs (hunk 2 above)
Comment (JoshGx2000) [reply frontend/src/cljs/imas_seamap/map/utils.cljs:231]: Yeah, I noticed that, but it has mucked up indentation, at least in my editor. I'll use them in future, knowing they are preferred.

File: frontend/src/cljs/imas_seamap/map/views.cljs (hunk 3 above)
Comment (JoshGx2000) [reply frontend/src/cljs/imas_seamap/map/views.cljs:211]: Once per state change, at least that's what my testing shows.

File: backend/catalogue/management/commands/generate_layer_previews.py (hunk 4 above)
Comment (JoshGx2000) [reply backend/catalogue/management/commands/generate_layer_previews.py:15]: Hmm, yeah. I thought I had directly committed this on the develop branch as part of a separate ticket. Not sure if I accidentally committed on this branch, or whether things got a bit confused when I merged develop into this branch.

File: frontend/src/cljs/imas_seamap/map/utils.cljs (hunk 1 above)
Comment (JoshGx2000) [reply frontend/src/cljs/imas_seamap/map/utils.cljs:213]: Yep. If it comes down to it I can always right extra code to merge the contents of all style elements (but this was the simplest solution).

**Pull request merged**